kubernetes/updater/pkg/controller/deployment.go

```diff
@@ -0,0 +1,127 @@
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	kclient "sigs.k8s.io/controller-runtime/pkg/client"
+	ctrllog "sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+// DeploymentVersionUpdater Reconciles a deployment by changing its img
+type DeploymentVersionUpdater struct {
+	VersionUpdater
+	kclient.Client
+	Scheme *runtime.Scheme
+}
+
+func (r *DeploymentVersionUpdater) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := ctrllog.FromContext(ctx).WithValues("namespacedname", req.NamespacedName)
+	ctx = ctrllog.IntoContext(ctx, log)
+
+	// Get the object we are reconciling
+	var obj appsv1.Deployment
+	if err := r.Get(ctx, req.NamespacedName, &obj); err != nil {
+		if apierrors.IsNotFound(err) {
+			// log.Info("not found")
+			return ctrl.Result{}, nil
+		}
+		// log.Error(err, "failed to get resource")
+		return ctrl.Result{}, trace.Wrap(err)
+	}
+
+	// Get the current and past version
+	currentVersion, err := getDeploymentVersion(&obj)
+	if err != nil {
+		switch trace.Unwrap(err).(type) {
+		case *trace.BadParameterError:
+			log.Info("Teleport container found, but failed to get version from the img tag. Will continue and do a version update.")
+		default:
+			return requeueLater, trace.Wrap(err)
+		}
+	}
+
+	image, err := r.Do(ctx, &obj, currentVersion)
+	if err != nil {
+		switch trace.Unwrap(err).(type) {
+		case *NoNewVersionError:
+			log.Info("Deployment is already up-to-date, not updating.")
+			return requeueLater, nil
+		case *MaintenanceNotTriggeredError:
+			log.Info("No maintenance triggered, not updating.")
+			return requeueLater, nil
+		case *trace.CompareFailedError:
+			log.Info("Image verification failed, not updating.")
+			return requeueLater, trace.Wrap(err)
+		default:
+			log.Error(err, "Unexpected error, not updating.")
+			return requeueLater, trace.Wrap(err)
+		}
+	}
+
+	err = setContainerImageFromPod(&obj.Spec.Template.Spec, teleportContainerName, image.String())
+	if err != nil {
+		return requeueLater, trace.Wrap(err)
+	}
+
+	err = r.Update(ctx, &obj)
+	if err != nil {
+		return requeueNow, trace.Wrap(err)
+	}
+	return requeueLater, nil
+}
+
+func (r *DeploymentVersionUpdater) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&appsv1.Deployment{}).
+		Complete(r)
+}
+
+func getDeploymentVersion(deployment *appsv1.Deployment) (string, error) {
+	var current string
+	image, err := getContainerImageFromPod(deployment.Spec.Template.Spec, teleportContainerName)
+	if err != nil {
+		return current, trace.Wrap(err)
+	}
+
+	// TODO: put this in a function and reuse for statefulset
+	imageRef, err := reference.ParseNamed(image)
+	if err != nil {
+		return current, trace.Wrap(err)
+	}
+	taggedImageRef, ok := imageRef.(reference.Tagged)
+	if !ok {
+		return "", trace.BadParameter("imageRef %s is not tagged", imageRef)
+	}
+	current = "v" + taggedImageRef.Tag()
```

Comment:
Check if v is already present. Currently, we do not include the v's, but in the future that might change

Reply:
I'll have the updater support reading version from images with tags prefixed with `v`. However, if we do the change, it will break the updater later in the "set version" phase as it will set the image tag without the `v`.